staking/sprmaker.go

```diff
@@ -0,0 +1,25 @@
+package staking
+
+import (
+	"context"
+	"github.com/pegnet/pegnet/spr"
+	"github.com/zpatrick/go-config"
+)
+
+type ISPRMaker interface {
+	NewSPR(ctx context.Context, dbht int32, config *config.Config) (*spr.StakingPriceRecord, error)
+}
+
+// SPRMaker
+// TODO: Should we change this at all?
+type SPRMaker struct {
```

Comment:
why empty struct?

Reply:
Empty struct struct{} is realized in special way in Go.

1) It’s a smallest building block in Go. It’s size is literally 0 bytes.
2) If has zero size. you may create a slice of 1000’s empty structures and this slice will be very tiny. Because really Go stores only a number of them in the slice but not them itself. The same story with channels.
3) All pointers to it always point to the same special place in memory.
4) Very useful In channels when you have notify about some event but you don’t need to pass any information about it, only a fact. Best solution is to pass an empty structure because it will only increment a counter in the channel but not assign memory, copy elements end so on. Sometime people use Boolean values for this purpose, but it’s much worse.
5) Zero size container for methods. You may want have a mock for testing interfaces. often you don’t need data on it just methods with predefined input and output.
6) Go has no Set object. Bit can be easily realized as a map[keyType]struct{}. This way map keeps only keys and no values.

Reply:
In this code, empty structs behave just like any other type, it works as method receivers. @sigrlami